### ISSUE TITLE
Kustomize command to add environment variables to containers in a kustomization

### ISSUE THREAD

**Issue by guibirow** [kubernetes-sigs/kustomize#2301]
Some applications allow configuration to be defined via Environment Variables. When deploying the yaml file with customize, we need to define these environment variables into the deployment.yaml file and pass the values accordingly.

In many cases, when the values are optional, we shouldn't need to provide any values or define the variables into the deployment.

Would make much more convenient if we could set these environment variable using customize commands instead of filling the deployment.yaml with all possible parameters.

Example:
Instead of defining the file like this:
```
kind: Deployment
apiVersion: apps/v1
kind: Deployment
metadata:
  name: wordpress
spec:
  template:
    spec:
      containers:
      - name: wordpress
        env:           // Just one for simplicity, this could have a long list of optional variables 
        - name: WORDPRESS_DB_HOST
          value: $(MYSQL_SERVICE)
```
We could define like this:

```
kind: Deployment
apiVersion: apps/v1
kind: Deployment
metadata:
  name: wordpress
spec:
  template:
    spec:
      containers:
      - name: wordpress
```

And inject environment variables with:

Specific container:
```
kustomize edit set env WORDPRESS_DB_HOST=$(MYSQL_SERVICE) container=wordpress
```
for specific deployment 
```
kustomize edit set env WORDPRESS_DB_HOST=$(MYSQL_SERVICE) deployment=wordpress
```
for a specific container in a deployment:
```
kustomize edit set env WORDPRESS_DB_HOST=$(MYSQL_SERVICE) deployment=wordpress container=wordpress
```
For all containers in the kustomization
```
kustomize edit set env WORDPRESS_DB_HOST=$(MYSQL_SERVICE)
```

This will be useful to provide Environment Variables to containers dynamically at deployment time without setup the yaml file in advance.

**Comment by omninonsense** [comment]
@guibirow Hi! Do would and [`ConfigMapGenerator`](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/plugins/builtins.md#configmapgenerator) and [`SecretGenerator`](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/plugins/builtins.md#secretgenerator) not suffice here? If so, why not?

**Comment by guibirow** [comment]
Probably yes, 
I had a look at it but couldn't identify how would it work to create the configMap or secret keys at deployment time using a value available only during deployment(e.g: environment variables or output from other commands)

This is why I suggested this approach, given the simplicity.

Is there any command similar to this to create thes configMaps?

**Comment by omninonsense** [comment]
It's not very well documented (if at all?), but the `KvLoader` used by both generators defers to the execution context's environment while parsing `env` files when the variable name isn't being set, so if you have an env file with this contents ([see here](https://github.com/kubernetes-sigs/kustomize/blob/85e9127071be589066b8a97083f92c2d991b04d7/api/kv/kv.go#L173-L175))

~~~dotenv
FOO=lalala
WORDPRESS_DB_HOST
~~~

the `WORDPRESS_DB_HOST` variable would be read from the environment by kustomize.

**Comment by guibirow** [comment]
Nice one, I will try that, thanks.

The only downside is the keys still need to be defined in advance and the names must match the keys throughout, but is already a good start.

Also seems is a better idea then setting the environment variables in the containers.

I will reformulate the issue, would still be useful to have a command line to create these keys at deployment time, so we can adjust the naming according to our needs.

**Comment by omninonsense** [comment]
> The only downside is the keys still need to be defined in advance and the names must match the keys throughout, but is already a good start.

From rereading your example it's a bit unclear what the effect of running it would be. Can you post an example kustomization and the desired outcome after running your command? I think I misunderstood it initially. 

~~~
kustomize edit set env WORDPRESS_DB_HOST=$(MYSQL_SERVICE) container=wordpress
~~~

Kustomize doesn't provide any editing capabilities, so this confuses me a little. What would be edited? The files themselves? Kustomize doesn't do that, AFAIK, but there's also no need for that, if all you need is renaming.

You can't avoid defining the names in advance at _some_ point. It'll have to be a constant name somewhere down chain, if even in code. I personally don't find the ability to rename/alias or define them via `kustomize` or `kubectl` that useful or practical. To me, it makes the situation more confusing, because it makes it harder to "follow" what's what or what's coming from where.

If this is _really_ necessary for some reason—situation that comes to mind is no control over the sourcode (in this case WordPress), and no control over the original names of the variables (let's say they're already used by other stuff in the CI pipeline and they're provided by the CI), the aliasing could still be done at the environment level, for example:

~~~sh
export WORDPRESS_DB_HOST=$(MYSQL_SERVICE)
kustomize build wordpress-deployment.yaml | kubectl apply -f -
~~~

or if you prefer to inline it:

~~~sh
WORDPRESS_DB_HOST=$(MYSQL_SERVICE) kustomize build wordpress-deployment.yaml | kubectl apply -f -
~~~

**Comment by guibirow** [comment]
Just an update on this issue.
I ended adding tokens to the file and doing `envsubst`
Much simpler and avoid the need to map the env-var to the same name as the container env var names.

So I can do like below
```
FOO=$(MY_FOO)
WORDPRESS_DB_HOST=$(MY_WP_DB)
```

**Comment by HariSekhon** [comment]
This would have been useful to be able to add say annotation to CI deployments via this field:

```
kubernetes.io/change-cause="deployed build $hashref"
```

Otherwise the `kubectl rollout history` has blank change-cause `<none>`, which would be useful if you want to track things in the kubectl CLI when deployments happened, or roll back quickly via the CLI.

Ideally this would be even better if the change-cause could capture the GMT timestamp and a templated / environment variable message

**Comment by afirth** [comment]
dup of #388, see also https://github.com/kubernetes-sigs/cli-experimental/issues/66

**Comment by boarder7395** [comment]
For others reading this:

Im not sure if this is the intention but I have found this would work:

export WORDPRESS_DB_HOST=$(MYSQL_SERVICE)
kustomize build ./wordpress | kubectl apply -f -

except the environment variable cannot have underscores other wise the replacement does not happen. So instead use:
env.yaml
____
FOO=lalala
WORDPRESSDBHOST

___
export WORDPRESSDBHOST=$(MYSQL_SERVICE)
kustomize build ./wordpress | kubectl apply -f -

Also pay specific attention to the lack of equal after WORDPRESSDBHOST because if an equal exists it will not pull from the environment variables.

**Comment by boarder7395** [comment]
Update on the above:

I was actually using kfctl with kustomize when I ran into the underscore issue with environment variables. Using kubectl kustomize, or kustomize directly does not look to have this issue. So disregard the above unless you are using kfctl to build your kustomize manifests.

**Comment by afirth** [comment]
this is finally documented in https://github.com/kubernetes/website/pull/30348

**Comment by joebowbeer** [comment]
> this is finally documented in [kubernetes/website#30348](https://github.com/kubernetes/website/pull/30348)

And [reverted](https://github.com/kubernetes/website/pull/35522) eight months later

**Comment by confiq** [comment]
they have good point, 

> Other tools (sed, jinja, erb, envsubst, kafka, helm, ksonnet, etc.) provide varying degrees of unstructured editting and/or embedded languages, and can be used instead of, or in a pipe with, kustomize. If you want to go all-in on configuration as a language, consider [cue](https://cuelang.org/).

> kustomize is going to stick to YAML in / YAML out.

https://kubectl.docs.kubernetes.io/faq/kustomize/eschewedfeatures/#build-time-side-effects-from-cli-args-or-env-variables

and I was looking for a way to get out of the YAML DRY principle... 

At this point, I'm really looking for better alternatives like cdk8s.io so I won't be sticking with YAML

**Comment by cyprianbergoniatmo** [comment]
> > this is finally documented in [kubernetes/website#30348](https://github.com/kubernetes/website/pull/30348)
> 
> but then [reverted](https://github.com/kubernetes/website/pull/35522) eight months later to fix [kubernetes/website#35669](https://github.com/kubernetes/website/issues/35669)

If you look at the source code, https://github.com/kubernetes-sigs/kustomize/blob/master/api/kv/kv.go#L163-L169 
It looks like that was a bug that will be deprecated
